diff -uNr a/component/isp2/sensor/aptina_ar0331/ar0331_cmos.c b/component/isp2/sensor/aptina_ar0331/ar0331_cmos.c
--- a/component/isp2/sensor/aptina_ar0331/ar0331_cmos.c	1970-01-01 08:00:00.000000000 +0800
+++ b/component/isp2/sensor/aptina_ar0331/ar0331_cmos.c	2016-05-21 01:02:07.812566022 +0800
@@ -0,0 +1,711 @@
+#if !defined(__AR0331_CMOS_H_)
+#define __AR0331_CMOS_H_
+
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include "hi_comm_sns.h"
+#include "hi_sns_ctrl.h"
+#include "mpi_isp.h"
+#include "mpi_ae.h"
+#include "mpi_awb.h"
+#include "mpi_af.h"
+
+#ifdef __cplusplus
+#if __cplusplus
+extern "C"{
+#endif
+#endif /* End of #ifdef __cplusplus */
+
+#define AR0331_ID 331
+
+/*set Frame End Update Mode 2 with HI_MPI_ISP_SetAEAttr and set this value 1 to avoid flicker in antiflicker mode */
+/*when use Frame End Update Mode 2, the speed of i2c will affect whole system's performance                       */
+/*increase I2C_DFT_RATE in Hii2c.c to 400000 to increase the speed of i2c                                         */
+#define CMOS_AR0331_ISP_WRITE_SENSOR_ENABLE (1)
+/****************************************************************************
+ * local variables                                                            *
+ ****************************************************************************/
+
+extern const unsigned int sensor_i2c_addr;
+extern unsigned int sensor_addr_byte;
+extern unsigned int sensor_data_byte;
+
+HI_U32 gu32FullLinesStd = 1100;
+HI_U8 gu8SensorMode = 0;
+
+#if CMOS_AR0331_ISP_WRITE_SENSOR_ENABLE
+ISP_SNS_REGS_INFO_S g_stSnsRegsInfo = {0};
+#endif
+
+static AWB_CCM_S g_stAwbCcm =
+{
+	4900,
+	{
+		0x0206, 0x80a0, 0x8066,
+		0x8063, 0x0190, 0x802d,
+		0x800f, 0x80bd, 0x01cc
+	},
+
+	3100,
+	{
+		0x01e2, 0x8097, 0x804b,
+		0x806d, 0x0185, 0x8018,
+		0x8010, 0x80f3, 0x0203
+	},
+
+	2470,
+	{
+		0x01a5, 0x8066, 0x803f,
+		0x805f, 0x0153, 0xb,
+		0x8061, 0x80e8, 0x249
+	}
+
+};
+
+static AWB_AGC_TABLE_S g_stAwbAgcTable =
+{
+	/* bvalid */
+	1,
+
+	/* saturation */
+	{0x80,0x80,0x6C,0x48,0x44,0x40,0x3C,0x38}
+};
+
+static ISP_CMOS_AGC_TABLE_S g_stIspAgcTable =
+{
+	/* bvalid */
+	1,
+
+	/* sharpen_alt_d */
+	{0x50,0x48,0x40,0x38,0x34,0x30,0x28,0x28},
+
+	//sharpen_alt_ud
+	{0x90,0x88,0x80,0x70,0x60,0x50,0x40,0x40},
+
+	//snr_thresh
+	{0x13,0x19,0x20,0x26,0x2c,0x32,0x38,0x38},
+
+	/* demosaic_lum_thresh */
+	{0x60,0x60,0x80,0x80,0x80,0x80,0x80,0x80},
+
+	/* demosaic_np_offset */
+	{0x0,0xa,0x12,0x1a,0x20,0x28,0x30,0x30},
+
+	/* ge_strength */
+	{0x55,0x55,0x55,0x55,0x55,0x55,0x37,0x37}
+};
+
+static ISP_CMOS_NOISE_TABLE_S g_stIspNoiseTable =
+{
+	/* bvalid */
+	1,
+
+	/* nosie_profile_weight_lut */
+	{0,  0,  0,  0,  0,  0,  11, 15, 17, 19, 20, 21, 22, 22, 23, 24,
+		25, 25, 26, 26, 26, 27, 27, 27, 28, 28, 28, 29, 29, 29, 29, 29,
+		30, 30, 30, 30, 30, 31, 31, 31, 31, 31, 32, 32, 32, 32, 32, 32,
+		32, 33, 33, 33, 33, 33, 33, 33, 33, 33, 34, 34, 34, 34, 34, 34,
+		34, 34, 34, 34, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 36, 36,
+		36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 36, 37, 37,
+		37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 37, 38, 38,
+		38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38},
+
+	/* demosaic_weight_lut */
+	{0, 11, 15, 17, 19, 20, 21, 22, 23, 23, 24, 25, 25, 26, 26, 26,
+		27, 27, 27, 28, 28, 28, 29, 29, 29, 29, 29, 30, 30, 30, 30, 30,
+		31, 31, 31, 31, 31, 32, 32, 32, 32, 32, 32, 32, 33, 33, 33, 33,
+		33, 33, 33, 33, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 35,
+		35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 35, 36, 36, 36, 36, 36,
+		36, 36, 36, 36, 36, 36, 36, 36, 36, 37, 37, 37, 37, 37, 37, 37,
+		37, 37, 37, 37, 37, 37, 37, 37, 37, 38, 38, 38, 38, 38, 38, 38,
+		38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 38}
+};
+
+static ISP_CMOS_DEMOSAIC_S g_stIspDemosaic =
+{
+	/* bvalid */
+	1,
+
+	/*vh_slope*/
+	0xe9,
+
+	/*aa_slope*/
+	0xd3,
+
+	/*va_slope*/
+	0xcc,
+
+	/*uu_slope*/
+	0xc2,
+
+	/*sat_slope*/
+	0x5d,
+
+	/*ac_slope*/
+	0xcf,
+
+	/*vh_thresh*/
+	0,
+
+	/*aa_thresh*/
+	0,
+
+	/*va_thresh*/
+	0,
+
+	/*uu_thresh*/
+	0,
+
+	/*sat_thresh*/
+	0x171,
+
+	/*ac_thresh*/
+	0x1b3
+};
+
+static ISP_CMOS_GAMMA_S g_stIspGamma =
+{
+	/* bvalid */
+	1,
+
+#if 1
+	{0  ,120 ,220 ,310 ,390 ,470 ,540 ,610 ,670 ,730 ,786 ,842 ,894 ,944 ,994 ,1050,
+		1096,1138,1178,1218,1254,1280,1314,1346,1378,1408,1438,1467,1493,1519,1543,1568,
+		1592,1615,1638,1661,1683,1705,1726,1748,1769,1789,1810,1830,1849,1869,1888,1907,
+		1926,1945,1963,1981,1999,2017,2034,2052,2069,2086,2102,2119,2136,2152,2168,2184,
+		2200,2216,2231,2247,2262,2277,2292,2307,2322,2337,2351,2366,2380,2394,2408,2422,
+		2436,2450,2464,2477,2491,2504,2518,2531,2544,2557,2570,2583,2596,2609,2621,2634,
+		2646,2659,2671,2683,2696,2708,2720,2732,2744,2756,2767,2779,2791,2802,2814,2825,
+		2837,2848,2859,2871,2882,2893,2904,2915,2926,2937,2948,2959,2969,2980,2991,3001,
+		3012,3023,3033,3043,3054,3064,3074,3085,3095,3105,3115,3125,3135,3145,3155,3165,
+		3175,3185,3194,3204,3214,3224,3233,3243,3252,3262,3271,3281,3290,3300,3309,3318,
+		3327,3337,3346,3355,3364,3373,3382,3391,3400,3409,3418,3427,3436,3445,3454,3463,
+		3471,3480,3489,3498,3506,3515,3523,3532,3540,3549,3557,3566,3574,3583,3591,3600,
+		3608,3616,3624,3633,3641,3649,3657,3665,3674,3682,3690,3698,3706,3714,3722,3730,
+		3738,3746,3754,3762,3769,3777,3785,3793,3801,3808,3816,3824,3832,3839,3847,3855,
+		3862,3870,3877,3885,3892,3900,3907,3915,3922,3930,3937,3945,3952,3959,3967,3974,
+		3981,3989,3996,4003,4010,4018,4025,4032,4039,4046,4054,4061,4068,4075,4082,4089,4095}
+#else  /*higher  contrast*/
+	{0  , 54 , 106, 158, 209, 259, 308, 356, 403, 450, 495, 540, 584, 628, 670, 713,
+		754 ,795 , 835, 874, 913, 951, 989,1026,1062,1098,1133,1168,1203,1236,1270,1303,
+		1335,1367,1398,1429,1460,1490,1520,1549,1578,1607,1635,1663,1690,1717,1744,1770,
+		1796,1822,1848,1873,1897,1922,1946,1970,1993,2017,2040,2062,2085,2107,2129,2150,
+		2172,2193,2214,2235,2255,2275,2295,2315,2335,2354,2373,2392,2411,2429,2447,2465,
+		2483,2501,2519,2536,2553,2570,2587,2603,2620,2636,2652,2668,2684,2700,2715,2731,
+		2746,2761,2776,2790,2805,2819,2834,2848,2862,2876,2890,2903,2917,2930,2944,2957,
+		2970,2983,2996,3008,3021,3033,3046,3058,3070,3082,3094,3106,3118,3129,3141,3152,
+		3164,3175,3186,3197,3208,3219,3230,3240,3251,3262,3272,3282,3293,3303,3313,3323,
+		3333,3343,3352,3362,3372,3381,3391,3400,3410,3419,3428,3437,3446,3455,3464,3473,
+		3482,3490,3499,3508,3516,3525,3533,3541,3550,3558,3566,3574,3582,3590,3598,3606,
+		3614,3621,3629,3637,3644,3652,3660,3667,3674,3682,3689,3696,3703,3711,3718,3725,
+		3732,3739,3746,3752,3759,3766,3773,3779,3786,3793,3799,3806,3812,3819,3825,3831,
+		3838,3844,3850,3856,3863,3869,3875,3881,3887,3893,3899,3905,3910,3916,3922,3928,
+		3933,3939,3945,3950,3956,3962,3967,3973,3978,3983,3989,3994,3999,4005,4010,4015,
+		4020,4026,4031,4036,4041,4046,4051,4056,4061,4066,4071,4076,4081,4085,4090,4095,4095}
+#endif
+};
+
+static  HI_U32   analog_gain_table[29]={
+	128,131,136,140,145,152,157,163,170,177,
+	185,194,204,215,227,240,256,272,293,314,
+	341,372,409,455,512,584,682,819,1024
+};
+
+
+static  HI_U32   gain_table[29] =
+{
+	1024,1054,1096,1126,1167,1218,1259,1311,1362,1423,1484,1556,1638,1720,1822,1925,
+	2048,2181,2345,2519,2734,2979,3276,3645,4096,4679,5458,6553,8192
+};
+
+HI_U32 cmos_get_isp_default(ISP_CMOS_DEFAULT_S *pstDef)
+{
+	if (HI_NULL == pstDef)
+	{
+		printf("null pointer when get isp default value!\n");
+		return -1;
+	}
+
+	memset(pstDef, 0, sizeof(ISP_CMOS_DEFAULT_S));
+
+	pstDef->stComm.u8Rggb = 0x1;      //1: rggb
+	pstDef->stComm.u8BalanceFe = 0x1;
+
+	pstDef->stDenoise.u8SinterThresh = 0x23;
+	pstDef->stDenoise.u8NoiseProfile= 0x1;      //0: use default profile table; 1: use calibrated profile lut, the setting for nr0 and nr1 must be correct.
+	pstDef->stDenoise.u16Nr0 = 0x0;
+	pstDef->stDenoise.u16Nr1 = 546;
+
+	pstDef->stDrc.u8DrcBlack = 0x00;
+	pstDef->stDrc.u8DrcVs = 0x04;     // variance space
+	pstDef->stDrc.u8DrcVi = 0x01;     // variance intensity
+	pstDef->stDrc.u8DrcSm = 0x30;     // slope max
+	pstDef->stDrc.u16DrcWl = 0x4FF;    // white level
+
+	memcpy(&pstDef->stNoiseTbl, &g_stIspNoiseTable, sizeof(ISP_CMOS_NOISE_TABLE_S));
+	memcpy(&pstDef->stAgcTbl, &g_stIspAgcTable, sizeof(ISP_CMOS_AGC_TABLE_S));
+	memcpy(&pstDef->stDemosaic, &g_stIspDemosaic, sizeof(ISP_CMOS_DEMOSAIC_S));
+	memcpy(&pstDef->stGamma, &g_stIspGamma, sizeof(ISP_CMOS_GAMMA_S));
+
+	return 0;
+}
+
+HI_U32 cmos_get_isp_black_level(ISP_CMOS_BLACK_LEVEL_S *pstBlackLevel)
+{
+	HI_S32  i;
+
+	if (HI_NULL == pstBlackLevel)
+	{
+		printf("null pointer when get isp black level value!\n");
+		return -1;
+	}
+
+	/* Don't need to update black level when iso change */
+	pstBlackLevel->bUpdate = HI_FALSE;
+
+	for (i=0; i<4; i++)
+	{
+		pstBlackLevel->au16BlackLevel[i] = 0xA8;
+	}
+
+	return 0;
+}
+
+HI_VOID cmos_set_pixel_detect(HI_BOOL bEnable)
+{
+	if (bEnable) /* setup for ISP pixel calibration mode */
+	{
+		sensor_write_register(0x300A, 0x1A5E);	//5fps
+		sensor_write_register(0x3012, 0x1996);	//max exposure lines
+		sensor_write_register(0x3060, 0x1300);	//AG, Context A
+		sensor_write_register(0x305E, 0x0128);	//DG, Context A
+	}
+	else /* setup for ISP 'normal mode' */
+	{
+		sensor_write_register(0x300A, 0x465);	//30fps
+	}
+
+	return;
+}
+
+HI_VOID cmos_set_wdr_mode(HI_U8 u8Mode)
+{
+	switch(u8Mode)
+	{
+		//sensor mode 0
+		case 0:
+			gu8SensorMode = 0;
+			// TODO:
+			break;
+			//sensor mode 1
+		case 1:
+			gu8SensorMode = 1;
+			// TODO:
+			break;
+
+		default:
+			printf("NOT support this mode!\n");
+			return;
+			break;
+	}
+
+	return;
+}
+
+static HI_S32 cmos_get_ae_default(AE_SENSOR_DEFAULT_S *pstAeSnsDft)
+{
+	if (HI_NULL == pstAeSnsDft)
+	{
+		printf("null pointer when get ae default value!\n");
+		return -1;
+	}
+
+	gu32FullLinesStd = 1100;
+
+	pstAeSnsDft->au8HistThresh[0] = 0xd;
+	pstAeSnsDft->au8HistThresh[1] = 0x28;
+	pstAeSnsDft->au8HistThresh[2] = 0x60;
+	pstAeSnsDft->au8HistThresh[3] = 0x80;
+
+	pstAeSnsDft->u8AeCompensation = 0x40;
+
+	pstAeSnsDft->u32LinesPer500ms = gu32FullLinesStd * 30 / 2;
+	pstAeSnsDft->u32FullLinesStd = gu32FullLinesStd;
+	pstAeSnsDft->u32FlickerFreq = 0;//60*256;//50*256;
+
+	pstAeSnsDft->stIntTimeAccu.enAccuType = AE_ACCURACY_LINEAR;
+	pstAeSnsDft->stIntTimeAccu.f32Accuracy = 1;
+	pstAeSnsDft->u32MaxIntTime = gu32FullLinesStd - 2;
+	pstAeSnsDft->u32MinIntTime = 2;
+	pstAeSnsDft->u32MaxIntTimeTarget = 65535;
+	pstAeSnsDft->u32MinIntTimeTarget = 2;
+
+	pstAeSnsDft->stAgainAccu.enAccuType = AE_ACCURACY_TABLE;
+	pstAeSnsDft->stAgainAccu.f32Accuracy = 0.0078125;
+	pstAeSnsDft->u32MaxAgain = 8192;  /* 8/0.0078125= 1024 */
+	pstAeSnsDft->u32MinAgain = 1024;
+	pstAeSnsDft->u32MaxAgainTarget = 8192;
+	pstAeSnsDft->u32MinAgainTarget = 1024;
+
+	pstAeSnsDft->stDgainAccu.enAccuType = AE_ACCURACY_LINEAR;
+	pstAeSnsDft->stDgainAccu.f32Accuracy = 0.0078125;
+	pstAeSnsDft->u32MaxDgain = 2047;/* 16 / 0.0078125 = 2047;*/
+	pstAeSnsDft->u32MinDgain = 128;
+	pstAeSnsDft->u32MaxDgainTarget = 2047;
+	pstAeSnsDft->u32MinDgainTarget = 128;
+
+	pstAeSnsDft->u32ISPDgainShift = 8;
+	pstAeSnsDft->u32MaxISPDgainTarget = 4 << pstAeSnsDft->u32ISPDgainShift;
+	pstAeSnsDft->u32MinISPDgainTarget = 1 << pstAeSnsDft->u32ISPDgainShift;
+
+	return 0;
+}
+
+static HI_S32 cmos_get_sensor_max_resolution(ISP_CMOS_SENSOR_MAX_RESOLUTION *pstSensorMaxResolution)
+{
+	if (HI_NULL == pstSensorMaxResolution)
+	{
+		printf("null pointer when get sensor max resolution \n");
+		return -1;
+	}
+
+	memset(pstSensorMaxResolution, 0, sizeof(ISP_CMOS_SENSOR_MAX_RESOLUTION));
+
+	pstSensorMaxResolution->u32MaxWidth  = 1920;
+	pstSensorMaxResolution->u32MaxHeight = 1080;
+
+	return 0;
+}
+
+
+/* the function of sensor set fps */
+static HI_VOID cmos_fps_set(HI_U8 u8Fps, AE_SENSOR_DEFAULT_S *pstAeSnsDft)
+{
+	switch(u8Fps)
+	{
+	case 30:
+		// Change the frame rate via changing the vertical blanking
+		gu32FullLinesStd = 1100;
+		pstAeSnsDft->u32MaxIntTime = gu32FullLinesStd - 2;
+		pstAeSnsDft->u32LinesPer500ms = gu32FullLinesStd * 30 / 2;
+		sensor_write_register(0x300A, 0x465);
+		break;
+
+	case 25:
+		// Change the frame rate via changing the vertical blanking
+		gu32FullLinesStd = 1100;
+		pstAeSnsDft->u32MaxIntTime = gu32FullLinesStd;
+		pstAeSnsDft->u32LinesPer500ms = gu32FullLinesStd * 25 / 2;
+		sensor_write_register(0x300A, 0x546);
+		break;
+
+	default:
+		break;
+	}
+
+	pstAeSnsDft->u32FullLinesStd = gu32FullLinesStd;
+
+	return;
+}
+
+static HI_VOID cmos_slow_framerate_set(HI_U16 u16FullLines,
+		AE_SENSOR_DEFAULT_S *pstAeSnsDft)
+{
+	sensor_write_register(0x300A, u16FullLines);
+
+	pstAeSnsDft->u32MaxIntTime = u16FullLines - 2;
+
+	return;
+}
+
+static HI_VOID cmos_init_regs_info(HI_VOID)
+{
+#if CMOS_AR0331_ISP_WRITE_SENSOR_ENABLE
+	HI_S32 i;
+	static HI_BOOL bInit = HI_FALSE;
+
+	if (HI_FALSE == bInit)
+	{
+		g_stSnsRegsInfo.enSnsType = ISP_SNS_I2C_TYPE;
+		g_stSnsRegsInfo.u32RegNum = 3;
+		for (i=0; i<g_stSnsRegsInfo.u32RegNum; i++)
+		{
+			g_stSnsRegsInfo.astI2cData[i].u8DevAddr = sensor_i2c_addr;
+			g_stSnsRegsInfo.astI2cData[i].u32AddrByteNum = sensor_addr_byte;
+			g_stSnsRegsInfo.astI2cData[i].u32DataByteNum = sensor_data_byte;
+		}
+		g_stSnsRegsInfo.astI2cData[0].bDelayCfg = HI_FALSE;
+		g_stSnsRegsInfo.astI2cData[0].u32RegAddr = 0x3012;
+		g_stSnsRegsInfo.astI2cData[1].bDelayCfg = HI_FALSE;
+		g_stSnsRegsInfo.astI2cData[1].u32RegAddr = 0x3060;
+		g_stSnsRegsInfo.astI2cData[2].bDelayCfg = HI_FALSE;
+		g_stSnsRegsInfo.astI2cData[2].u32RegAddr = 0x305E;
+
+		g_stSnsRegsInfo.bDelayCfgIspDgain = HI_FALSE;
+
+		bInit = HI_TRUE;
+	}
+#endif
+	return;
+}
+
+
+
+/* while isp notify ae to update sensor regs, ae call these funcs. */
+static HI_VOID cmos_inttime_update(HI_U32 u32IntTime)
+{
+	static HI_U32 _last_exposure_time = 0xFFFFFFFF;
+
+	if(_last_exposure_time == u32IntTime)
+	{
+		return ;
+	}else
+	{
+		_last_exposure_time = u32IntTime;
+	}
+
+#if CMOS_AR0331_ISP_WRITE_SENSOR_ENABLE
+	cmos_init_regs_info();
+	g_stSnsRegsInfo.astI2cData[0].u32Data = u32IntTime;
+#else
+	sensor_write_register(0x3012, u32IntTime);
+#endif
+	return;
+}
+
+
+static HI_VOID cmos_again_calc_table(HI_U32 u32InTimes,AE_SENSOR_GAININFO_S *pstAeSnsGainInfo)
+{
+	int i;
+
+	if(HI_NULL == pstAeSnsGainInfo)
+	{
+		printf("null pointer when get ae sensor gain info  value!\n");
+		return;
+	}
+
+	pstAeSnsGainInfo->u32GainDb = 0;
+	pstAeSnsGainInfo->u32SnsTimes = 1024;
+
+	if (u32InTimes >= gain_table[28])
+	{
+		pstAeSnsGainInfo->u32SnsTimes = gain_table[28];
+		pstAeSnsGainInfo->u32GainDb = 28;
+		return ;
+	}
+
+	for(i = 1; i < 29; i++)
+	{
+		if(u32InTimes < gain_table[i])
+		{
+			pstAeSnsGainInfo->u32SnsTimes = gain_table[i - 1];
+			pstAeSnsGainInfo->u32GainDb = i - 1;
+			break;
+		}
+
+	}
+
+	return;
+
+}
+
+
+static HI_VOID cmos_gains_update(HI_U32 u32Again, HI_U32 u32Dgain)
+{
+	static HI_U32 _last_again = 0xFFFFFFFF;
+	static HI_U32 _last_dgain = 0xFFFFFFFF;
+	static HI_U32 _last_exposure = 0xFFFFFFFF;
+
+	if((_last_again ==  u32Again ) && (_last_dgain == u32Dgain)&& g_stSnsRegsInfo.astI2cData[0].u32Data == _last_exposure)
+	{
+		return ;
+	}
+	else
+	{
+		_last_again = u32Again;
+		_last_dgain= u32Dgain;
+		_last_exposure = g_stSnsRegsInfo.astI2cData[0].u32Data;
+	}
+
+#if CMOS_AR0331_ISP_WRITE_SENSOR_ENABLE
+	cmos_init_regs_info();
+
+	if( u32Again < 16)
+	{
+		g_stSnsRegsInfo.astI2cData[1].u32Data = 0x00 + u32Again;
+	}else if( u32Again <24)
+	{
+		g_stSnsRegsInfo.astI2cData[1].u32Data = 0x10 + (u32Again - 16) * 2;
+	}else if( u32Again <28)
+	{
+		g_stSnsRegsInfo.astI2cData[1].u32Data = 0x20 + (u32Again - 24) * 4;
+	}else if( u32Again == 28)
+	{
+		g_stSnsRegsInfo.astI2cData[1].u32Data = 0x30;
+	}
+
+	g_stSnsRegsInfo.astI2cData[2].u32Data = u32Dgain;
+
+
+	HI_MPI_ISP_SnsRegsCfg(&g_stSnsRegsInfo);
+#endif
+	return;
+}
+
+static HI_S32 cmos_get_awb_default(AWB_SENSOR_DEFAULT_S *pstAwbSnsDft)
+{
+	if (HI_NULL == pstAwbSnsDft)
+	{
+		printf("null pointer when get awb default value!\n");
+		return -1;
+	}
+
+	memset(pstAwbSnsDft, 0, sizeof(AWB_SENSOR_DEFAULT_S));
+
+	pstAwbSnsDft->u16WbRefTemp = 4900;
+
+	pstAwbSnsDft->au16GainOffset[0] = 0x198;
+	pstAwbSnsDft->au16GainOffset[1] = 0x100;
+	pstAwbSnsDft->au16GainOffset[2] = 0x100;
+	pstAwbSnsDft->au16GainOffset[3] = 0x1b2;
+
+	pstAwbSnsDft->as32WbPara[0] = 55;
+	pstAwbSnsDft->as32WbPara[1] = 24;
+	pstAwbSnsDft->as32WbPara[2] = -177;
+	pstAwbSnsDft->as32WbPara[3] = 271396;
+	pstAwbSnsDft->as32WbPara[4] = 128;
+	pstAwbSnsDft->as32WbPara[5] = -220863;
+
+	memcpy(&pstAwbSnsDft->stCcm, &g_stAwbCcm, sizeof(AWB_CCM_S));
+	memcpy(&pstAwbSnsDft->stAgcTbl, &g_stAwbAgcTable, sizeof(AWB_AGC_TABLE_S));
+
+	return 0;
+}
+
+HI_VOID sensor_global_init()
+{
+	gu8SensorMode = 0;
+}
+
+
+/****************************************************************************
+ * callback structure                                                       *
+ ****************************************************************************/
+HI_S32 cmos_init_sensor_exp_function(ISP_SENSOR_EXP_FUNC_S *pstSensorExpFunc)
+{
+	memset(pstSensorExpFunc, 0, sizeof(ISP_SENSOR_EXP_FUNC_S));
+
+	pstSensorExpFunc->pfn_cmos_sensor_init = sensor_init;
+	pstSensorExpFunc->pfn_cmos_sensor_global_init = sensor_global_init;
+	pstSensorExpFunc->pfn_cmos_get_isp_default = cmos_get_isp_default;
+	pstSensorExpFunc->pfn_cmos_get_isp_black_level = cmos_get_isp_black_level;
+	pstSensorExpFunc->pfn_cmos_set_pixel_detect = cmos_set_pixel_detect;
+	pstSensorExpFunc->pfn_cmos_set_wdr_mode = cmos_set_wdr_mode;
+	pstSensorExpFunc->pfn_cmos_get_sensor_max_resolution = cmos_get_sensor_max_resolution;
+
+	return 0;
+}
+
+HI_S32 cmos_init_ae_exp_function(AE_SENSOR_EXP_FUNC_S *pstExpFuncs)
+{
+	memset(pstExpFuncs, 0, sizeof(AE_SENSOR_EXP_FUNC_S));
+
+	pstExpFuncs->pfn_cmos_get_ae_default = cmos_get_ae_default;
+	pstExpFuncs->pfn_cmos_fps_set = cmos_fps_set;
+	pstExpFuncs->pfn_cmos_slow_framerate_set= cmos_slow_framerate_set;
+	pstExpFuncs->pfn_cmos_inttime_update = cmos_inttime_update;
+	pstExpFuncs->pfn_cmos_gains_update = cmos_gains_update;
+	pstExpFuncs->pfn_cmos_again_calc_table  = cmos_again_calc_table;
+
+	return 0;
+}
+
+HI_S32 cmos_init_awb_exp_function(AWB_SENSOR_EXP_FUNC_S *pstExpFuncs)
+{
+	memset(pstExpFuncs, 0, sizeof(AWB_SENSOR_EXP_FUNC_S));
+
+	pstExpFuncs->pfn_cmos_get_awb_default = cmos_get_awb_default;
+
+	return 0;
+}
+
+int sensor_register_callback(void)
+{
+	HI_S32 s32Ret;
+	ALG_LIB_S stLib;
+	ISP_SENSOR_REGISTER_S stIspRegister;
+	AE_SENSOR_REGISTER_S  stAeRegister;
+	AWB_SENSOR_REGISTER_S stAwbRegister;
+
+	cmos_init_sensor_exp_function(&stIspRegister.stSnsExp);
+	s32Ret = HI_MPI_ISP_SensorRegCallBack(AR0331_ID, &stIspRegister);
+	if (s32Ret)
+	{
+		printf("sensor register callback function failed!\n");
+		return s32Ret;
+	}
+
+	stLib.s32Id = 0;
+	strcpy(stLib.acLibName, HI_AE_LIB_NAME);
+	cmos_init_ae_exp_function(&stAeRegister.stSnsExp);
+	s32Ret = HI_MPI_AE_SensorRegCallBack(&stLib, AR0331_ID, &stAeRegister);
+	if (s32Ret)
+	{
+		printf("sensor register callback function to ae lib failed!\n");
+		return s32Ret;
+	}
+
+	stLib.s32Id = 0;
+	strcpy(stLib.acLibName, HI_AWB_LIB_NAME);
+	cmos_init_awb_exp_function(&stAwbRegister.stSnsExp);
+	s32Ret = HI_MPI_AWB_SensorRegCallBack(&stLib, AR0331_ID, &stAwbRegister);
+	if (s32Ret)
+	{
+		printf("sensor register callback function to ae lib failed!\n");
+		return s32Ret;
+	}
+
+	return 0;
+}
+
+int sensor_unregister_callback(void)
+{
+	HI_S32 s32Ret;
+	ALG_LIB_S stLib;
+
+	s32Ret = HI_MPI_ISP_SensorUnRegCallBack(AR0331_ID);
+	if (s32Ret)
+	{
+		printf("sensor unregister callback function failed!\n");
+		return s32Ret;
+	}
+
+	stLib.s32Id = 0;
+	strcpy(stLib.acLibName, HI_AE_LIB_NAME);
+	s32Ret = HI_MPI_AE_SensorUnRegCallBack(&stLib, AR0331_ID);
+	if (s32Ret)
+	{
+		printf("sensor unregister callback function to ae lib failed!\n");
+		return s32Ret;
+	}
+
+	stLib.s32Id = 0;
+	strcpy(stLib.acLibName, HI_AWB_LIB_NAME);
+	s32Ret = HI_MPI_AWB_SensorUnRegCallBack(&stLib, AR0331_ID);
+	if (s32Ret)
+	{
+		printf("sensor unregister callback function to ae lib failed!\n");
+		return s32Ret;
+	}
+
+	return 0;
+}
+
+#ifdef __cplusplus
+#if __cplusplus
+}
+#endif
+#endif /* End of #ifdef __cplusplus */
+
+#endif // __AR0331_CMOS_H_
diff -uNr a/component/isp2/sensor/aptina_ar0331/ar0331_sensor_ctl.c b/component/isp2/sensor/aptina_ar0331/ar0331_sensor_ctl.c
--- a/component/isp2/sensor/aptina_ar0331/ar0331_sensor_ctl.c	1970-01-01 08:00:00.000000000 +0800
+++ b/component/isp2/sensor/aptina_ar0331/ar0331_sensor_ctl.c	2016-05-21 00:54:29.246602807 +0800
@@ -0,0 +1,320 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+
+#ifdef HI_GPIO_I2C
+#include "gpioi2c_ex.h"
+#include "gpio_i2c.h"
+#else
+#include "hi_i2c.h"
+#endif
+
+const unsigned char sensor_i2c_addr = 0x20;   /* I2C Address of AR0330 */
+const unsigned int  sensor_addr_byte = 2;     /* ADDR byte of AR0330 */
+const unsigned int  sensor_data_byte = 2;     /* DATA byte of AR0330 */
+
+int sensor_read_register(int addr)
+{
+	// TODO:
+
+	return 0;
+}
+
+
+int sensor_write_register(int addr, int data)
+{
+#ifdef HI_GPIO_I2C
+	int fd = -1;
+	int ret;
+	I2C_DATA_S  i2c_data;
+
+	fd = open("/dev/gpioi2c_ex", 0);
+	if(fd<0)
+	{
+		printf("Open gpioi2c_ex error!\n");
+		return -1;
+	}
+
+	i2c_data.dev_addr = sensor_i2c_addr;
+	i2c_data.reg_addr = addr;
+	i2c_data.addr_byte_num = sensor_addr_byte;
+	i2c_data.data = data;
+	i2c_data.data_byte_num = sensor_data_byte;
+
+	ret = ioctl(fd, GPIO_I2C_WRITE, &i2c_data);
+
+	if (ret)
+	{
+		printf("GPIO-I2C write faild!\n");
+		return -1;
+	}
+
+	close(fd);
+#else
+
+	int fd = -1;
+	int ret;
+	I2C_DATA_S i2c_data;
+
+	fd = open("/dev/hi_i2c", 0);
+	if(fd < 0)
+	{
+		printf("Open i2c device error!\n");
+		return -1;
+	}
+
+	i2c_data.dev_addr = sensor_i2c_addr;
+	i2c_data.reg_addr = addr;
+	i2c_data.addr_byte_num = sensor_addr_byte;
+	i2c_data.data = data;
+	i2c_data.data_byte_num = sensor_data_byte;
+
+	ret = ioctl(fd,  CMD_I2C_WRITE, &i2c_data);
+	if(ret)
+	{
+		printf("i2c write failed!\n");
+		return -1 ;
+	}
+
+	close(fd);
+#endif
+	return 0;
+}
+
+static void delay_ms(int ms) {
+	usleep(ms*1000);
+}
+
+void sensor_init()
+{
+	/* Reset */
+	sensor_write_register(0x301A, 0x10D9);	/* reset_register */
+	delay_ms(100);
+	sensor_write_register(0x301A, 0x10D8);	/* reset_register */
+	delay_ms(200);
+
+	/* AR0331 Rev1 Linear sequencer load - 1.3 */
+	sensor_write_register(0x3088, 0x8000);	/* seq_ctrl_port */
+	sensor_write_register(0x3086, 0x4A03);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0443);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1645);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4045);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x6017);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5045);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x404B);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1244);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x6134);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4A31);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4342);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4560);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2714);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3DFF);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3DFF);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3DEA);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2704);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3D10);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2705);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3D10);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2715);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3527);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x053D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1045);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4027);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0427);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x143D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFF3D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFF3D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xEA62);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2728);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x3627);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x083D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x6444);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2C2C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2C2C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4B01);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x432D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4643);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1647);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x435F);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4F50);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2604);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2684);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2027);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFC53);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0D5C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0D57);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5417);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0955);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5649);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5307);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5303);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4D28);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x6C4C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0928);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2C28);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x294E);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5C09);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4500);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4580);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x26A6);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x27F8);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1702);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x27FA);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5C0B);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1718);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x26A2);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5C03);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1744);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x27F2);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1702);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2809);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1710);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1628);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x084D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1A26);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x8316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1627);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFA45);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xA017);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0727);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFB17);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2945);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x8017);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0827);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xFA17);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x285D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x170E);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2681);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5301);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1740);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5302);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1710);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2683);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2682);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x484D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4E28);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x094C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0B17);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5F27);	/* seq_data_port */
+	sensor_write_register(0x3086, 0xF217);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1428);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0816);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4D1A);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1616);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x27FA);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2603);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5C01);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4540);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2798);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x172A);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4A0A);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0B43);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x279C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4560);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1707);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x279D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1725);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4540);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1708);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2798);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5D53);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0D26);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x455C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x014B);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1244);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5251);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1702);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x6018);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4A03);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x0443);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1658);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5943);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x165A);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4316);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x5B43);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4540);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x279C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4560);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1707);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x279D);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1725);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x4540);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1710);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2798);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1720);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x224B);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x1244);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2C2C);	/* seq_data_port */
+	sensor_write_register(0x3086, 0x2C2C);	/* seq_data_port */
+
+	sensor_write_register(0x301A, 0x10D8);	/* reset_register */
+	sensor_write_register(0x30B0, 0x0000);	/* digital_test */
+	sensor_write_register(0x31AC, 0x0C0C);	/* data_format_bits */
+
+	/* PLL settings - Parallel */
+	sensor_write_register(0x302A, 0x0008);	/* vt_pix_clk_div */
+	sensor_write_register(0x302C, 0x0001);	/* vt_sys_clk_div */
+	sensor_write_register(0x302E, 0x0002);	/* pre_pll_clk_div */
+	sensor_write_register(0x3030, 0x002C);	/* pll_multiplier */
+	sensor_write_register(0x3036, 0x000C);	/* op_pix_clk_div */
+	sensor_write_register(0x3038, 0x0001);	/* op_sys_clk_div */
+
+	/* Linear Mode HD 1080P30 - Parallel */
+	sensor_write_register(0x3002, 0x00E4);	/* y_addr_start */
+	sensor_write_register(0x3004, 0x0042);	/* x_addr_start */
+	sensor_write_register(0x3006, 0x0523);	/* y_addr_end */
+	sensor_write_register(0x3008, 0x07C9);	/* x_addr_end */
+	sensor_write_register(0x300A, 0x0465);	/* frame_length_lines */
+	sensor_write_register(0x300C, 0x044C);	/* line_length_pck */
+	sensor_write_register(0x3012, 0x0416);	/* coarse_integration_time */
+	sensor_write_register(0x3014, 0x0000);	/* fine_integration_time */
+	sensor_write_register(0x30A2, 0x0001);	/* x_odd_inc */
+	sensor_write_register(0x30A6, 0x0001);	/* y_odd_inc */
+	sensor_write_register(0x3040, 0x0000);	/* read_mode */
+	sensor_write_register(0x31AE, 0x0301);	/* serial_format */
+	/* Linear Mode */
+	sensor_write_register(0x3082, 0x0009);	/* operation_mode_ctrl */
+	sensor_write_register(0x30BA, 0x06EC);	/* digital_ctrl */
+	/* 2D motion compensation OFF */
+	sensor_write_register(0x318C, 0x0000);	/* hdr_mc_ctrl2 */
+	/* ALTM Bypassed */
+	sensor_write_register(0x2400, 0x0003);	/* altm_control */
+	/* ADACD Disabled */
+	sensor_write_register(0x3200, 0x0000);	/* adacd_control */
+
+	sensor_write_register(0x31D0, 0x0000);	/* companding */
+	sensor_write_register(0x30FE, 0x0080);	/* noise_pedestal */
+	sensor_write_register(0x320A, 0x0080);	/* adacd_pedestal */
+	sensor_write_register(0x2450, 0x0000);	/* altm_out_pedestal */
+	sensor_write_register(0x301E, 0x00A8);	/* data_pedestal */
+
+	/* Linear Mode Devware Color Setup */
+
+	sensor_write_register(0x3012, 0x02A0);	/* coarse_integration_time */
+	sensor_write_register(0x3064, 0x1982);	/* smia_test */
+
+	/* Analog Settings */
+	sensor_write_register(0x2440, 0x0002);	/* altm_control_damper */
+	sensor_write_register(0x30F4, 0x4000);	/* -- */
+	sensor_write_register(0x3180, 0x8029);	/* delta_dk_control */
+	sensor_write_register(0x3198, 0x061E);	/* hdr_mc_ctrl8 */
+	sensor_write_register(0x3ED2, 0x3F46);	/* dac_ld_6_7 */
+	sensor_write_register(0x3ED4, 0x8F6C);
+	sensor_write_register(0x3ED6, 0x6666);
+	sensor_write_register(0x3EDA, 0x8899);
+	sensor_write_register(0x3EE6, 0x00F0);
+
+	sensor_write_register(0x301A, 0x10DC);
+
+	printf("Aptina AR0331 sensor 1080p 30fps init success!\n");
+}
+
diff -uNr a/component/isp2/sensor/aptina_ar0331/Makefile b/component/isp2/sensor/aptina_ar0331/Makefile
--- a/component/isp2/sensor/aptina_ar0331/Makefile	1970-01-01 08:00:00.000000000 +0800
+++ b/component/isp2/sensor/aptina_ar0331/Makefile	2016-05-21 00:54:30.074597327 +0800
@@ -0,0 +1,65 @@
+#
+# sensor lib Makefile
+#
+
+ifeq ($(PARAM_FILE), )
+	PARAM_FILE:=../../../../Makefile.param
+	include $(PARAM_FILE)
+endif
+
+ISP_PATH := $(MPP_PATH)/component/isp2
+EXT_PATH := $(MPP_PATH)/$(EXTDRV)
+3A_PATH  := $(ISP_PATH)/3a
+LIBPATH = $(ISP_PATH)/lib
+OBJPATH = ./obj
+
+ifeq ($(CONFIG_GPIO_I2C),Y)
+HI_GPIO_I2C:=HI_GPIO_I2C
+else
+HI_GPIO_I2C:=HI_GPIO_XXX
+endif
+
+ARFLAGS = rcv
+ARFLAGS_SO = -shared -fPIC -o
+CFLAGS  = -Wall -fPIC -D$(HI_FPGA) -D$(HI_GPIO_I2C)
+CFLAGS  += -O2
+
+ifeq ($(HIGDB),HI_GDB)
+CFLAGS += -g
+endif
+BUS_DIR := $(EXT_PATH)/ssp-sony/
+
+ISP_INC := $(ISP_PATH)/include
+3A_INC := $(3A_PATH)/include
+INC := -I$(BUS_DIR) -I$(REL_INC) -I$(ISP_INC) -I$(3A_INC)
+
+ifeq ($(CONFIG_GPIO_I2C), Y)
+    INC += -I$(EXT_PATH)/gpio-i2c-ex
+    INC += -I$(EXT_PATH)/gpio-i2c
+else
+    INC += -I$(EXT_PATH)/i2c
+    INC += -I$(EXT_PATH)/hi_i2c
+endif
+
+COMPILE = $(CC) $(CFLAGS) $(DFLAGS) -lm
+
+$(OBJPATH)/%.o: ./%.c
+	@[ -e $(LIBPATH) ] || mkdir $(LIBPATH)
+	@[ -e $(OBJPATH) ] || mkdir $(OBJPATH)
+	@$(COMPILE) -o $@ -c $< $(INC)
+
+SRCS = $(wildcard ./*.c)
+OBJS = $(SRCS:%.c=%.o)
+OBJS := $(OBJS:./%=obj/%)
+
+TARGETLIB := $(LIBPATH)/libsns_ar0331_1080p.a
+TARGETLIB_SO := $(LIBPATH)/libsns_ar0331_1080p.so
+
+all:$(TARGETLIB)
+$(TARGETLIB):$(OBJS)
+	@($(AR) $(ARFLAGS) $(TARGETLIB) $(OBJS))
+	@($(CC) $(ARFLAGS_SO) $(TARGETLIB_SO) $(OBJS))
+
+clean:
+	@$(RM) -rf $(TARGETLIB) $(OBJS)
+	@$(RM) -rf $(LIBPATH) $(OBJPATH)
