diff -uNr mpp/Makefile.param mpp.modified/Makefile.param
--- mpp/Makefile.param	2012-11-28 18:49:06.000000000 +0800
+++ mpp.modified/Makefile.param	2015-12-10 18:38:34.583810325 +0800
@@ -122,13 +122,13 @@
 export CC:=$(CROSS)gcc
 export AR:=$(CROSS)ar
 
-export AUDIO_PATH?=$(SDK_PATH)/mpp/audiolib/$(CROSS)
-export JPEGD_PATH?=$(SDK_PATH)/mpp/jpegdlib
+export AUDIO_PATH?=$(MPP_PATH)/audiolib/$(CROSS)
+export JPEGD_PATH?=$(MPP_PATH)/jpegdlib
 
 # Define the object output directory 
-export REL_INC := $(SDK_PATH)/mpp/include_$(HIARCH)
-export REL_LIB := $(SDK_PATH)/mpp/lib_$(HIARCH)
-export REL_KO  := $(SDK_PATH)/mpp/ko_$(HIARCH)
+export REL_INC := $(MPP_PATH)/include_$(HIARCH)
+export REL_LIB := $(MPP_PATH)/lib_$(HIARCH)
+export REL_KO  := $(MPP_PATH)/ko_$(HIARCH)
 
 # Define the lib of audio codec
 export AUDIO_LIBA := $(REL_LIB)/libVoiceEngine.a	\
diff -uNr mpp/sample/common/loadbmp.c mpp.modified/sample/common/loadbmp.c
--- mpp/sample/common/loadbmp.c	2012-11-28 18:49:06.000000000 +0800
+++ mpp.modified/sample/common/loadbmp.c	2015-12-10 16:02:49.743212873 +0800
@@ -13,7 +13,7 @@
                                                     {0, 8, 8, 8},   /*RGB888*/
                                                     {8, 8, 8, 8}    /*ARGB8888*/
                                                     };
-inline HI_U16 OSD_MAKECOLOR_U16(HI_U8 r, HI_U8 g, HI_U8 b, OSD_COMP_INFO compinfo)
+static inline HI_U16 OSD_MAKECOLOR_U16(HI_U8 r, HI_U8 g, HI_U8 b, OSD_COMP_INFO compinfo)
 {
     HI_U8 r1, g1, b1;
     HI_U16 pixel = 0;
diff -uNr mpp/sample/common/sample_comm_audio.c mpp.modified/sample/common/sample_comm_audio.c
--- mpp/sample/common/sample_comm_audio.c	2012-11-28 18:49:06.000000000 +0800
+++ mpp.modified/sample/common/sample_comm_audio.c	2015-12-10 16:02:12.668522924 +0800
@@ -1083,7 +1083,7 @@
     return HI_SUCCESS;
 }
 
-inline HI_S32 SAMPLE_COMM_AUDIO_StopHdmi(HI_VOID)
+static inline HI_S32 SAMPLE_COMM_AUDIO_StopHdmi(HI_VOID)
 {
     HI_S32 s32Ret;
     VO_DEV VoDev = 0;
diff -uNr mpp/sample/hi3520A/Makefile.param mpp.modified/sample/hi3520A/Makefile.param
--- mpp/sample/hi3520A/Makefile.param	2012-11-28 18:49:06.000000000 +0800
+++ mpp.modified/sample/hi3520A/Makefile.param	2015-12-10 18:39:41.788234151 +0800
@@ -1,14 +1,14 @@
 # Hisilicon Hi3516 sample Makefile.param
 
 ifeq ($(PARAM_FILE), )
-     PARAM_FILE:=../../../../mpp/Makefile.param
+     PARAM_FILE:=../../../Makefile.param
      include $(PARAM_FILE)
 endif
 
 INC_FLAGS := -I$(REL_INC)\
-		-I$(SDK_PATH)/mpp/sample/common
+		-I$(MPP_PATH)/sample/common
 
-INC_FLAGS += -I$(REL_INC) -I$(SDK_PATH)/mpp/$(EXTDRV)/tw2865 -I$(SDK_PATH)/mpp/$(EXTDRV)/tw2960 -I$(SDK_PATH)/mpp/$(EXTDRV)/tlv320aic31
+INC_FLAGS += -I$(REL_INC) -I$(MPP_PATH)/$(EXTDRV)/tw2865 -I$(MPP_PATH)/$(EXTDRV)/tw2960 -I$(MPP_PATH)/$(EXTDRV)/tlv320aic31
 
 CFLAGS := -Wall -g $(INC_FLAGS) -D$(HIARCH) -DHICHIP=$(HICHIP) -DSENSOR_TYPE=$(SENSOR_TYPE) -D$(HIDBG) -D$(HI_FPGA) -lpthread -lm
 
@@ -32,5 +32,5 @@
 CXXFlAGS +=-march=armv7-a -mcpu=cortex-a9 -mfloat-abi=softfp -mfpu=vfpv3-d16
 endif
 
-COMM_SRC := $(wildcard $(SDK_PATH)/mpp/sample/common/*.c)
+COMM_SRC := $(wildcard $(MPP_PATH)/sample/common/*.c)
 COMM_OBJ := $(COMM_SRC:%.c=%.o)
