From 00ebfc2f24b66290b352e6316574ba5c764799cb Mon Sep 17 00:00:00 2001
From: Watson Xu <xuhuashan@gmail.com>
Date: Wed, 12 Jul 2017 23:42:30 +0800
Subject: the buf argument for write should be double quoted.

---
 src/pipe.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/pipe.cpp b/src/pipe.cpp
index 01211b3..d0572c7 100644
--- a/src/pipe.cpp
+++ b/src/pipe.cpp
@@ -83,5 +83,5 @@ ssize_t Pipe::read(void *buffer, unsigned int &nbytes)
 void Pipe::signal()
 {
   // TODO: ignoring return of read/write generates warning; maybe relevant for eventloop work...
-  ::write(_fd_write, '\0', 1);
+  ::write(_fd_write, "\0", 1);
 }
-- 
2.13.0

