diff -uNr a/src/message.cpp b/src/message.cpp
--- a/src/message.cpp	2011-11-28 23:03:19.000000000 +0800
+++ b/src/message.cpp	2016-09-25 11:07:23.230932915 +0800
@@ -350,7 +350,8 @@
       (
         (DBusMessageIter *) & (to._iter),
         from.type(),
-        from.type() == DBUS_TYPE_VARIANT ? NULL : sig,
+        ((from.type() == DBUS_TYPE_VARIANT) || \
+         (from.type() == DBUS_TYPE_ARRAY)) ? sig : NULL,
         (DBusMessageIter *) & (to_container._iter)
       );
 
